contents/handbook/engineering/async-migrations.md

```diff
@@ -0,0 +1,222 @@
+
+<details>
+
+<summary><b>Example async migration</b></summary>
+
+<br />
+
+
+
+```python
+from posthog.constants import AnalyticsDBMS
+from posthog.special_migrations.definition import SpecialMigrationDefinition, SpecialMigrationOperation
+from posthog.version_requirement import ServiceVersionRequirement
+from ee.clickhouse.client import sync_execute
+
+class Migration(SpecialMigrationDefinition):
+
+    description = "Migrate table_x to a new schema"
+
+    # minimum posthog version for running this migration
+    posthog_min_version = "1.28.0"
+
+    # maximum posthog version this migration can be run on - it becomes a requirement from 1.31.1 onwards
+    posthog_max_version = "1.31.0"
+
+    # if clickhouse is on a version below 21.6.0, this migration won't run
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0"),
+    ]
+
+    # the special migration that needs to have completed before this one
+    depends_on = "0003_previous_special_migration"
+
+    # ordered list of operations in the migration
+    operations = [
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.POSTGRES,
+            sql="CREATE TABLE table_x_new ( key VARCHAR, value VARCHAR ) ENGINE = ReplacingMergeTree('value') ORDER BY key",
+            rollback="DROP TABLE table_x_new",
+        ),
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.POSTGRES,
+            sql="INSERT INTO table_x_new (col1) SELECT key, value FROM table_x",
+            rollback="TRUNCATE TABLE table_x_new",
+        ),
+    ]
+    
+    # only run if the table doesn't already exist
+    def is_required(self):
+        result = sync_execute("SELECT count(*) FROM system.tables WHERE database='posthog' AND name='table_x_new'")
+        return result[0][0] == 0 
+
+    # check if we're not running low on storage
+    def healthcheck(self):
+        result = sync_execute("SELECT free_space FROM system.disks")
+        if result[0][0] < 1000000000:
+            return (False, "ClickHouse available storage below 1GB") 
+
+        return (True, None)
+
+    # update progress based on the total rows moved
+    def progress(self, _):
+        result = sync_execute(f"SELECT count(*) FROM table_x")
+        result2 = sync_execute(f"SELECT count(*) FROM table_x_new")
+        total_rows_to_move = result2[0][0]
+        total_rows_moved = result[0][0]
+
+        progress = 100 * total_rows_moved / total_rows_to_move
+        return progress
+
+```
+
+</details>
+
+
+### Workflow and architecture 
+
+#### Setup
+
+When the Django server boots up - a setup step for async migrations happens, which does the following:
+
+1. Imports all the migration definitions 
+2. Populates a dependencies map and in-memory record of migration definitions
+3. Creates a database record for each
+4. Check if all async migrations necessary for this PostHog version have completed (else don't start)
```

Comment:
As I called out in the PR this is a problem for the upgrade path specifically (https://github.com/PostHog/posthog/pull/7446#discussion_r765339085), we need to fail earlier here and not let folks get into a state where they tried an upgrade and errored out on special migrations. Please explain how we're going to solve that.

on that note - there are also customers who kubectl apply stuff (we don't really need to support them as that's very custom), but how can we make their experience best?